src/Microsoft.DocAsCode.Build.TableOfContents/BuildTocDocument.cs

```diff
@@ -105,11 +111,130 @@ private void BuildCore(TocItemViewModel item, FileModel model, IHostService host
             }
         }
 
-        private static TocItemViewModel RestructureTableOfContent(TocItemViewModel toc, IHostService host)
+        private void RestructureCore(TocItemViewModel item, List<TocItemViewModel> parent, IList<TreeItemRestructure> restructures)
+        {
+            if (item.Items != null && item.Items.Count > 0)
+            {
+                var parentItems = new List<TocItemViewModel>(item.Items);
+                foreach (var i in item.Items)
+                {
+                    RestructureCore(i, parentItems, restructures);
+                }
+
+                item.Items = new TocViewModel(parentItems);
+            }
+
+            foreach (var restruction in restructures)
+            {
+                if (Matches(item, restruction))
+                {
+                    RestructureItem(item, parent, restruction);
+                }
+            }
+        }
+
+        private void RestructureItem(TocItemViewModel item, List<TocItemViewModel> parentItems, TreeItemRestructure restruction)
+        {
+            var index = parentItems.IndexOf(item);
+            if (index < 0)
+            {
+                Logger.LogWarning($"Unable to find {restruction.Key}, it is probably removed or replaced by other restructions.");
```

Comment:
can hit with multiple restructures, test case TestTocRestructureWithRestructureConflicts covers the scenario

---
In reply to: [97011166](https://github.com/dotnet/docfx/pull/1204#discussion_r97011166) [](ancestors = 97011166)